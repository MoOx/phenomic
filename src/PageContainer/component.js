// @flow
/* eslint-disable react/sort-comp */
import React, { Component, PropTypes } from "react"
import { findDOMNode } from "react-dom"

import urlify from "../_utils/urlify"

type DefaultProps = {
  defaultLayout: string,
}

type Props = {
  pages: Object,
  params: {
    splat: string,
  },
  layouts: Object,
  defaultLayout: string,
  getPage: Function,
  setPageNotFound: Function,
  logger: Object,
}

type Context = {
  collection: PhenomicCollection,
}

// react-router does not return leading and trailing slashes
// so we need to normalize according to collection data
const splatToUrl = (string) => ("/" + urlify(string))

const isDevelopment = (): boolean => process.env.NODE_ENV !== "production"
const isClient = (): boolean => typeof window !== "undefined"
const isDevelopmentClient = (): boolean => isDevelopment() && isClient()

let catchLinks
let browserHistory

if (isClient()) {
  catchLinks = require("../_utils/catch-links").default
  browserHistory = require("../client").browserHistory
}

function find(
  collection: PhenomicCollection,
  pageUrl: string
): Object {
  return collection.find((item) => (
    item.__url === pageUrl ||
    item.__url === pageUrl + "/"||
    item.__resourceUrl === pageUrl
  ))
}

function getBase(location: Object): string {
  return (
    location.protocol + "//" + location.host +
    process.env.PHENOMIC_USER_PATHNAME
  )
}

function adjustCurrentUrl(location: Object, item: Object, props: Props): void {
  // adjust url (eg: missing trailing slash)
  const currentExactPageUrl = location.href
    .replace(getBase(location), "/")
    .replace(location.hash, "")

  if (currentExactPageUrl !== item.__url) {
    props.logger.info(
      "phenomic: PageContainer: " +
      `replacing by '${ currentExactPageUrl }' to '${ item.__url }'`
    )
    if (browserHistory) {
      browserHistory.replace(item.__url)
    }
  }
}

function getLayout(
  layout: string, props: Props
): ReactClass<any> | void {
  if (props.layouts && props.layouts[layout]) {
    return props.layouts[layout]
  }
}

class PageContainer extends Component<DefaultProps, Props, void> {
  _content: HTMLElement;

  propTypes: Props;

  static contextTypes = {
    collection: PropTypes.arrayOf(PropTypes.object),
    layouts: PropTypes.object,
  };

  static defaultProps: DefaultProps = {
    layouts: {},
    defaultLayout: "Page",
    logger: console,
  };

  constructor(props: Props) {
    super(props)

    if (!getLayout(props.defaultLayout, props)) {
      props.logger.error(
        "phenomic: PageContainer: " +
        `default layout "${ props.defaultLayout }" not provided. `
      )
    }
  }

  componentWillMount() {
    this.preparePage(this.props, this.context)
  }

  componentDidMount() {
    this.catchInternalLink()
  }

  componentWillReceiveProps(nextProps: Props): void {
    this.preparePage(nextProps, this.context)
  }

  componentDidUpdate() {
    this.catchInternalLink()
  }

  catchInternalLink() {
    if (!isClient()) {
      return
    }

    if (this._content) {
      const layoutDOMElement = findDOMNode(this._content)
      if (layoutDOMElement) {
        let bodyContainers =
          Array.prototype.slice.call(
            layoutDOMElement.querySelectorAll(".phenomic-BodyContainer")
          )
        if (!bodyContainers.length) {
          bodyContainers = [ layoutDOMElement ]
        }
        catchLinks(bodyContainers, (href) => {
          const pageUrl = href.replace(process.env.PHENOMIC_USER_PATHNAME, "/")
          if (!find(this.context.collection, pageUrl)) {
            return false
          }
          if (browserHistory) {
            browserHistory.push(pageUrl)
          }
          return true
        })
      }
    }
  }

  preparePage(props: Props, context: Context): void {
    const pageUrl = splatToUrl(props.params.splat)
    if (isDevelopmentClient()) {
      props.logger.info(
        `phenomic: PageContainer: '${ pageUrl }' rendering...`
      )
    }

    const item = find(context.collection, pageUrl)
    if (isClient() && item) {
      adjustCurrentUrl(window.location, item, props)
    }

    const page = props.pages[pageUrl]
    if (!page) {
      if (item) {
        props.getPage(item.__url, item.__dataUrl)
      }
      else {
        props.logger.error(
          `phenomic: PageContainer: ${ pageUrl } is a page not found.`
        )
        props.setPageNotFound(pageUrl)
      }
    }
    else {
      if (page.error) {
        return
      }

      const Layout = getLayout(page.type, props)
      if (page.type !== undefined && !Layout) {
        props.logger.error(
          "phenomic: PageContainer: " +
          `Unkown page type: "${ page.type }" component not available in ` +
          "\"layouts\" property. " +
          `Please check the "layout" or "type" of page "${ page }" header.`
        )
      }
    }
  }

  saveContentRef: Function = (ref: HTMLElement): void => {
    this._content = ref
  };

  render() {
    const { props } = this

    const pageUrl = splatToUrl(props.params.splat)
    const page = props.pages[pageUrl]

    if (!page) {
      if (isDevelopmentClient()) {
        props.logger.info(`phenomic: PageContainer: '${ pageUrl }' no data`)
      }
      return null
    }
    if (isDevelopmentClient()) {
      props.logger.info(`phenomic: PageContainer: '${ pageUrl }'`, page)
    }

    if (
      typeof page !== "object" ||
      page.toString() !== "[object Object]"
    ) {
      props.logger.info(
        `phenomic: PageContainer: page ${ pageUrl } should be an object`
      )
      return null
    }

    const PageLoading = getLayout("PageLoading", props)
    const PageError = getLayout("PageError", props)
    const LayoutFallback = getLayout(props.defaultLayout, props)
    const Layout = getLayout(page.type, props) || LayoutFallback

    return (
      <div>
        {
          !page.error && page.loading && PageLoading &&
          <PageLoading />
        }
        {
          !!page.error && !PageError &&
          <div style={ { "text-align": "center" } }>
            <h1>{ page.error }</h1>
            <p>{ page.errorText }</p>
          </div>
        }
        {
          !!page.error && PageError &&
          <PageError { ...page } />
        }
        {
          !page.error && !page.loading && Layout &&
          <Layout ref={ this.saveContentRef } { ...page } />
        }
      </div>
    )
  }
}

export default PageContainer
