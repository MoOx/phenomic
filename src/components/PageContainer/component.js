// @flow

/* eslint-disable react/sort-comp */
import React, { Component, PropTypes } from "react"
import { findDOMNode } from "react-dom"

import urlify from "../../_utils/urlify"
import catchLinks from "../../_utils/catch-links"
import { browserHistory } from "../../client"

const logPrefix = "phenomic: PageContainer:"

type DefaultProps = {
  defaultLayout: string,
}

type Props = {
  pages: Object,
  params: {
    splat: string,
  },
  layouts: Object,
  defaultLayout: string,
  getPage: Function,
  setPageNotFound: Function,
  logger: Object,
}

type Context = {
  collection: PhenomicCollection,
}

// react-router does not return leading and trailing slashes
// so we need to normalize according to collection data
const splatToUrl = (string) => {
  const url = "/" + urlify(string)
  return (url === "//") ? "/" : url
}

function find(
  collection: PhenomicCollection,
  pageUrl: string
): ?PhenomicCollectionItem {
  return collection.find((item) => (
    item.__url === pageUrl ||
    item.__url === pageUrl + "/" ||
    item.__resourceUrl === pageUrl
  ))
}

function getBase(location: Object): string {
  return (
    location.protocol + "//" + location.host +
    process.env.PHENOMIC_USER_PATHNAME
  )
}

function adjustCurrentUrl(location: Object, item: Object, props: Props): void {
  // adjust url (eg: missing trailing slash)
  const currentExactPageUrl = location.href.replace(getBase(location), "/")
  const itemURL = item.__url + location.search + location.hash

  if (currentExactPageUrl !== itemURL) {
    props.logger.info(
      `${ logPrefix } replacing by '${ currentExactPageUrl }' to '${ itemURL }'`
    )
    if (browserHistory) {
      browserHistory.replace(itemURL)
    }
  }
}

function getLayout(
  layout: string, props: Props
): ReactClass<any> | void {
  if (props.layouts && props.layouts[layout]) {
    return props.layouts[layout]
  }
}

class PageContainer extends Component<DefaultProps, Props, void> {
  _content: HTMLElement;

  propTypes: Props;

  static contextTypes = {
    collection: PropTypes.arrayOf(PropTypes.object),
    layouts: PropTypes.object,
  };

  static defaultProps: DefaultProps = {
    layouts: {},
    defaultLayout: "Page",
    logger: console,
  };

  componentWillMount() {
    const { props } = this
    if (!getLayout(props.defaultLayout, props)) {
      props.logger.error(
        `${ logPrefix } default layout "${ props.defaultLayout }" not provided.`
      )
    }
    this.preparePage(this.props, this.context)
  }

  componentDidMount() {
    this.catchInternalLink()
  }

  componentWillReceiveProps(nextProps: Props): void {
    this.preparePage(nextProps, this.context)
  }

  componentDidUpdate() {
    this.catchInternalLink()
  }

  catchInternalLink() {
    const layoutDOMElement = findDOMNode(this)

    if (layoutDOMElement) {
      let bodyContainers =
        Array.prototype.slice.call(
          layoutDOMElement.querySelectorAll(".phenomic-BodyContainer")
        )
      if (!bodyContainers.length) {
        bodyContainers = [ layoutDOMElement ]
      }
      catchLinks(bodyContainers, (href) => {
        const pageUrl = href.replace(process.env.PHENOMIC_USER_PATHNAME, "/")
        if (!find(this.context.collection, pageUrl)) {
          return false
        }
        if (browserHistory) {
          browserHistory.push(pageUrl)
        }
        return true
      })
    }
  }

  preparePage(props: Props, context: Context): void {
    const pageUrl = splatToUrl(props.params.splat)
    if (process.env.NODE_ENV !== "production") {
      props.logger.info(
        `${ logPrefix } '${ pageUrl }' rendering...`
      )
    }

    const item = find(context.collection, pageUrl)
    if (typeof window !== "undefined" && item) {
      adjustCurrentUrl(window.location, item, props)
    }

    const page = props.pages[pageUrl]
    if (!page) {
      if (item) {
        props.getPage(item.__url, item.__dataUrl)
      }
      else {
        props.logger.error(
          `${ logPrefix } ${ pageUrl } is a page not found.`
        )
        props.setPageNotFound(pageUrl)
      }
    }
    else {
      if (page.error) {
        return
      }

      const Layout = getLayout(page.type, props)
      if (page.type !== undefined && !Layout) {
        props.logger.error(
          `${ logPrefix } Unkown page type: "${ page.type }"` +
          "component not available in \"layouts\" property." +
          `Please check the "layout" or "type" of page "${ page }" header.`
        )
      }
    }
  }

  render() {
    const { props } = this
    const { collection } = this.context

    const pageUrl = splatToUrl(props.params.splat)
    // page url from redux store
    const page = props.pages[pageUrl]
    const partialPageHead = collection.find((pageData) => {
      return pageUrl === pageData.__url
    }) || {}

    if (!page) {
      if (process.env.NODE_ENV !== "production") {
        props.logger.info(`${ logPrefix } '${ pageUrl }' no data`)
      }
      // return null
    }
    if (process.env.NODE_ENV !== "production") {
      props.logger.info(`${ logPrefix } '${ pageUrl }'`, page)
    }

    if (
      typeof page !== "object" ||
      page.toString() !== "[object Object]"
    ) {
      props.logger.info(
        `${ logPrefix } page ${ pageUrl } should be an object`
      )
      return null
    }
    const PageLoading = getLayout("PageLoading", props)
    const PageError = getLayout("PageError", props)
    const LayoutFallback = getLayout(props.defaultLayout, props)
    const Layout = getLayout(
      (
        partialPageHead.type || partialPageHead.layout ||
        // page.type is head type||layout too
        page.type
      ),
      props
    ) || LayoutFallback

    if (page.error) {
      if (!PageError) {
        return (
          <div style={{ "text-align": "center" }}>
            <h1>{ page.error }</h1>
            <p>{ page.errorText }</p>
          </div>
        )
      }
      return <PageError { ...page } />
    }
    else if (
      page.isLoading &&
      PageLoading &&
      (Layout && !Layout.hasLoadingState)
    ) {
      return <PageLoading />
    }
    else if (Layout) {
      return (
        <Layout
          // head will be overwritten by "page"
          // (since page contains a head when loaded)
          head={ partialPageHead }
          { ...page }
        />
      )
    }
    }

    return null
  }
}

export default PageContainer
