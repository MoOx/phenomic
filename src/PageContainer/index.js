// deprecated
import colors from "chalk"

import PageContainer from "../components/PageContainer"

console.log("⚠️ " + colors.yellow(
  "'phenomic/lib/PageContainer' reference is deprecated.\n" +
  "Please use `import { PageContainer } from \"phenomic\" instead`."
))

export default PageContainer
