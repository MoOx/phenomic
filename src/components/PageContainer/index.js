import { connect } from "react-redux"

// eslint-disable-next-line import/no-namespace
import * as pageActions from "../../redux/modules/pages"

import PageContainer from "./component"

export default connect(
  ({ pages }) => {
    return { pages }
  },
  (dispatch) => {
    return {
      getPage: (...args) => dispatch(pageActions.get(...args)),
      setPageNotFound: (...args) => dispatch(pageActions.setNotFound(...args)),
    }
  },
)(PageContainer)
