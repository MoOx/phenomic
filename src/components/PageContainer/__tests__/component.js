import React, { createElement as jsx } from "react"
import { createRenderer } from "react-addons-test-utils"

import PageContainer from "../component"

// fixtures
/* eslint-disable react/no-multi-comp */
const noop = () => {}
const Page = () => <div className="Page" />
const PageError = () => <div className="PageError" />
const AnotherPage = () => <div className="AnotherPage" />

it("should render a Page if page is ok", () => {
  const renderer = createRenderer()
  renderer.render(
    jsx(
      PageContainer,
      {
        params: { splat: "" },
        pages: { "/": {} },
        getPage: noop,
        setPageNotFound: noop,
        layouts: { Page },
      }
    ),
    {
      collection: [],
    },
  )
  expect(renderer.getRenderOutput()).toMatchSnapshot()
})

it("should try to get a page if no page in cache", () => {
  const renderer = createRenderer()
  const getPage = jest.fn()
  const setPageNotFound = jest.fn()

  renderer.render(
    jsx(
      PageContainer,
      {
        params: { splat: "" },
        pages: { },
        getPage,
        setPageNotFound,
        layouts: { Page },
        logger: {
          info: () => {},
        },
      }
    ),
    {
      collection: [
        {
          __url: "/",
          __dataUrl: "/j.son",
        },
      ],
    },
  )
  renderer.getRenderOutput()
  expect(getPage).toBeCalledWith("/", "/j.son")
  expect(setPageNotFound).not.toBeCalled()
})

it(`should render a visible error if page is not ok and no PageError
available`, () => {
  const renderer = createRenderer()
  renderer.render(
    jsx(
      PageContainer,
      {
        params: { splat: "" },
        pages: { "/": { error: "Test", errorText: "" } },
        getPage: noop,
        setPageNotFound: noop,
        layouts: { Page },
      }
    ),
    {
      collection: [],
    },
  )

  expect(renderer.getRenderOutput()).toMatchSnapshot()
})

it("should render PageError if page not found and PageError is available",
() => {
  const renderer = createRenderer()
  renderer.render(
    jsx(
      PageContainer,
      {
        params: { splat: "" },
        pages: { "/": { error: "Test" } },
        getPage: noop,
        setPageNotFound: noop,
        layouts: { Page, PageError },
      }
    ),
    {
      collection: [],
    },
  )

  expect(renderer.getRenderOutput()).toMatchSnapshot()
})

it("should render a another page layout if defaultLayout is used", () => {
  const renderer = createRenderer()
  renderer.render(
    jsx(
      PageContainer,
      {
        params: { splat: "" },
        pages: { "/": {} },
        getPage: noop,
        setPageNotFound: noop,
        defaultLayout: "AnotherPage",
        layouts: { AnotherPage },
      }
    ),
    {
      collection: [],
    },
  )

  expect(renderer.getRenderOutput()).toMatchSnapshot()
})
