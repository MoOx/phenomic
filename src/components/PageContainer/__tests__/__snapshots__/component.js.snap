exports[`test should render PageError if page not found and PageError is available 1`] = `
<function PageError() /* istanbul ignore next */{return (/* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */"div", /* istanbul ignore next */{ className: "PageError" }));}
  error="Test" />
`;

exports[`test should render a Page if page is ok 1`] = `<function Page() /* istanbul ignore next */{return (/* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */"div", /* istanbul ignore next */{ className: "Page" }));} />`;

exports[`test should render a another page layout if defaultLayout is used 1`] = `<function AnotherPage() /* istanbul ignore next */{return (/* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */"div", /* istanbul ignore next */{ className: "AnotherPage" }));} />`;

exports[`test should render a visible error if page is not ok and no PageError
available 1`] = `
<div
  style={
    Object {
      "text-align": "center",
    }
  }>
  <h1>
    Test
  </h1>
  <p />
</div>
`;
